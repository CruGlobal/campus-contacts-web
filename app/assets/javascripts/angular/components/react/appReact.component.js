import template from './appReact.html';

angular.module('missionhubApp').component('appReact', {
    controller: AppReactController,
    template: template,
});

function AppReactController($rootScope) {}
