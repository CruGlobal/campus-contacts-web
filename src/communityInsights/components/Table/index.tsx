import React, { useState } from 'react';
import { useQuery } from 'react-apollo-hooks';
import { useTranslation } from 'react-i18next';
import styled from '@emotion/styled';
import { withTheme } from 'emotion-theming';

import leftArrow from '../../assets/icons/arrow-left.svg';
import leftArrowActive from '../../assets/icons/arrow-left-active.svg';
import rightArrow from '../../assets/icons/arrow-right.svg';
import rightArrowActive from '../../assets/icons/arrow-right-active.svg';

const LoadingContainer = styled.div`
    text-align: center;
    padding: 25px;
`;

const TableContainer = styled.table`
    table-layout: fixed;
    width: 100%;
`;

const Header = styled.thead`
    text-align: left;
`;

const Content = styled.tbody``;

const Column = styled.th`
    font-weight: normal;
    font-size: 14px;
    line-height: 60px;
    color: ${({ theme }) => theme.colors.secondary};
    padding: 0 15px;
`;

const Row = styled.tr`
    text-align: left;
    height: 64px;

    :nth-of-type(odd) {
        background-color: #eceef2;
    }
`;

const Cell = styled.td`
    font-weight: normal;
    font-size: 14px;
    line-height: 20px;
    color: ${({ theme }) => theme.colors.primary};
    padding: 0 15px;
    max-width: 200px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    &:first-of-type {
        padding: 0 25px;
    }
    &:last-child {
        padding: 0 5px;
    }
`;

const Pagination = styled.div`
    width: 100%;
    height: 64px;
    display: flex;
    align-items: center;
    justify-content: center;
`;

const LeftArrow = styled.div`
    height: 16px;
    width: 16px;
    background-size: contain;
    background-repeat: no-repeat;
    background-image: url(${leftArrow});
    margin-right: 20px;

    &.active {
        cursor: pointer;
        background-image: url(${leftArrowActive});
    }
`;

const RightArrow = styled.div`
    height: 16px;
    width: 16px;
    background-size: contain;
    background-repeat: no-repeat;
    background-image: url(${rightArrow});
    margin-left: 20px;

    &.active {
        cursor: pointer;
        background-image: url(${rightArrowActive});
    }
`;

const Pages = styled.div`
    display: flex;
`;
const Page = styled.div`
    width: 32px;
    height: 32px;
    font-size: 14px;
    line-height: 20px;
    color: ${({ theme }) => theme.colors.secondary};
    align-items: center;
    display: flex;
    justify-content: center;

    &:not(:last-child) {
        margin-right: 4px;
    }

    &.active {
        border-radius: 16px;
        background-color: ${({ theme }) => theme.colors.highlight};
        color: #ffffff;
    }
`;

const PAGE_SIZE = 5;
const PAGE_NUMBERS_SIZE = 5;

interface Props {
    query: any;
    headers: any;
    mapRows: (data: any) => any;
    mapPage: (data: any) => any;
    variables: any;
}

const Table = ({ query, headers, mapRows, mapPage, variables }: Props) => {
    const [pageNumber, setPageNumber] = useState(1);

    const { data, loading, refetch } = useQuery(query, {
        variables: {
            ...variables,
            first: PAGE_SIZE,
        },
    });
    const { t } = useTranslation('insights');

    if (loading) {
        return <LoadingContainer>{t('loading')}</LoadingContainer>;
    }

    const rows = mapRows(data);
    const page = mapPage(data);

    const nextPage = () => {
        if (page.hasNextPage) {
            refetch({
                first: PAGE_SIZE,
                last: undefined,
                after: page.endCursor,
                before: undefined,
            });
            setPageNumber(pageNumber + 1);
        }
    };

    const previousPage = () => {
        if (page.hasPreviousPage) {
            refetch({
                first: undefined,
                last: PAGE_SIZE,
                after: undefined,
                before: page.startCursor,
            });
            setPageNumber(pageNumber > 1 ? pageNumber - 1 : 0);
        }
    };

    const renderPages = () => {
        const pages = [];
        const pageIndex = Math.floor(PAGE_NUMBERS_SIZE / 2);
        const startIndex =
            pageNumber - pageIndex > 0 ? pageNumber - pageIndex : 1;
        for (let i = startIndex; i < startIndex + PAGE_NUMBERS_SIZE; i++) {
            pages.push(
                <Page key={i} className={i === pageNumber ? 'active' : ''}>
                    {i}
                </Page>,
            );
        }
        return pages;
    };

    return (
        <div>
            <TableContainer>
                <Header>
                    <tr>
                        {headers.map((header: string, index: string) => (
                            <Column key={index}>{header}</Column>
                        ))}
                    </tr>
                </Header>
                <Content>
                    {rows.map((row: any, index: string) => (
                        <Row key={index}>
                            {row.map((cell: string, index: string) => (
                                <Cell key={index}>{cell}</Cell>
                            ))}
                        </Row>
                    ))}
                </Content>
            </TableContainer>
            <Pagination>
                <LeftArrow
                    className={page.hasPreviousPage ? 'active' : ''}
                    onClick={previousPage}
                />
                <Pages>{renderPages()}</Pages>
                <RightArrow
                    className={page.hasNextPage ? 'active' : ''}
                    onClick={nextPage}
                />
            </Pagination>
        </div>
    );
};

export default withTheme(Table);
