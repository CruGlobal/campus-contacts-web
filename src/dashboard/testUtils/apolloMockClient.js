import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
import { buildClientSchema } from 'graphql/utilities';
import { createMockerFromIntrospection } from 'fraql/mock';
import gql from 'fraql';

import introspectionQuery from '../../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument, options) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
