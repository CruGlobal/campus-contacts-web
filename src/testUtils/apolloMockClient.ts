import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
// @ts-ignore
import { buildClientSchema } from 'graphql/utilities';
// @ts-ignore
import { createMockerFromIntrospection } from 'fraql/mock';
// @ts-ignore
import gql from 'fraql';

import introspectionQuery from '../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument: any, options: any) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
