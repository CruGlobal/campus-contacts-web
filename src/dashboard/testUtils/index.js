import React, { ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';

import { createApolloMockClient } from './apolloMockClient';
import defaultTheme from '../defaultTheme';
import { ThemeProvider } from 'emotion-theming';
import AppContext from '../appContext';

export const renderWithContext = (component, { mocks } = {}, context = {}) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={context}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });

    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
