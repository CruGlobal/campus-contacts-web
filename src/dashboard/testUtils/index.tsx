import React, { ReactNode } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

import defaultTheme from '../defaultTheme';
import AppContext from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

export const renderWithContext = (
    component: any,
    { mocks: mocks = {}, appContext = {} } = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    // @ts-ignore
    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
