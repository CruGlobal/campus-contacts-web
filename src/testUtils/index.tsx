import React, { ReactNode, ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { IMocks } from 'graphql-tools';

import defaultTheme from '../defaultTheme';
import {
    AppContext,
    AppContextValue,
    appContextDefaultValue,
} from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

interface RenderWithContextParams {
    mocks?: IMocks;
    appContext?: AppContextValue;
}

export const renderWithContext = (
    component: ReactElement,
    {
        mocks: mocks = {},
        appContext = appContextDefaultValue,
    }: RenderWithContextParams = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children?: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
