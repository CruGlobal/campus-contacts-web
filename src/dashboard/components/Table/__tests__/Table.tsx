import React from 'react';
import gql from 'graphql-tag';
import { waitForElement } from '@testing-library/react';

import { renderWithContext } from '../../../testUtils';
import Table from '../';

const QUERY = gql`
    query globalCommunityChallenges($first: Int, $after: String) {
        globalCommunityChallenges(first: $first, after: $after) {
            nodes {
                id
                title
                acceptedCount
                completedCount
                createdAt
                endDate
            }
            pageInfo {
                hasNextPage
                startCursor
                endCursor
                hasPreviousPage
            }
        }
    }
`;

describe('<Table />', () => {
    it('should render properly in loading', async () => {
        const mapRows = () => [['cell1', 'cell2'], ['cell3', 'cell4']];
        const mapPage = () => ({});

        renderWithContext(
            <Table
                query={QUERY}
                headers={['header-1', 'header-2', 'header-3']}
                mapRows={mapRows}
                mapPage={mapPage}
                variables={{}}
            />,
        ).snapshot();
    });

    it('should render properly with data', async () => {
        const mapRows = () => [['cell1', 'cell2'], ['cell3', 'cell4']];
        const mapPage = () => ({});

        const { snapshot, getByText } = renderWithContext(
            <Table
                query={QUERY}
                headers={['header-1', 'header-2', 'header-3']}
                mapRows={mapRows}
                mapPage={mapPage}
                variables={{}}
            />,
            {
                mocks: {
                    Query: () => ({
                        globalCommunityChallenges: () => ({
                            nodes: () => [],
                            pageInfo: () => {},
                        }),
                    }),
                },
                appContext: {
                    orgId: 1,
                },
            },
        );

        await waitForElement(() => getByText('header-1'));
        snapshot();
    });
});
