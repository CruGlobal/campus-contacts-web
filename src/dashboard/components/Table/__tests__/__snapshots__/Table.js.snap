// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render properly in loading 1`] = `
<div>
  <div
    class="css-1b8qkfw"
  >
    Loading...
  </div>
</div>
`;

exports[`<Table /> should render properly with data 1`] = `
<div>
  <div>
    <table
      class="css-1f3ydut"
    >
      <thead
        class="css-tfuwtm"
      >
        <tr>
          <th
            class="css-1rfzrj7"
          >
            header-1
          </th>
          <th
            class="css-1rfzrj7"
          >
            header-2
          </th>
          <th
            class="css-1rfzrj7"
          >
            header-3
          </th>
        </tr>
      </thead>
      <tbody
        class="css-1esthja"
      >
        <tr
          class="css-e9u13j"
        >
          <td
            class="css-10yiwq9"
          >
            cell1
          </td>
          <td
            class="css-10yiwq9"
          >
            cell2
          </td>
        </tr>
        <tr
          class="css-e9u13j"
        >
          <td
            class="css-10yiwq9"
          >
            cell3
          </td>
          <td
            class="css-10yiwq9"
          >
            cell4
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="css-elql06"
    >
      <div
        class=" css-14swvyq"
      />
      <div
        class="css-1mzzuk6"
      >
        <div
          class="active css-mjxiyf"
        >
          1
        </div>
        <div
          class=" css-mjxiyf"
        >
          2
        </div>
        <div
          class=" css-mjxiyf"
        >
          3
        </div>
        <div
          class=" css-mjxiyf"
        >
          4
        </div>
        <div
          class=" css-mjxiyf"
        >
          5
        </div>
      </div>
      <div
        class=" css-1pqqm8"
      />
    </div>
  </div>
</div>
`;
